packages/runtime/container-runtime/src/summaryGenerator.ts

```diff
@@ -14,12 +14,28 @@ import {
 } from "./summarizerTypes";
 import { IClientSummaryWatcher } from "./summaryCollection";
 
-export const checkNotTimeout = <T>(something: T | IPromiseTimerResult | undefined): something is T => {
-    if (something === undefined) {
-        return false;
-    }
-    return (something as IPromiseTimerResult).timerResult === undefined;
+export type RaceTimerResult<T> = {
+    readonly result: IPromiseTimerResult["timerResult"];
+} | {
+    readonly result: "timeoutError" | "error";
+    readonly error: any;
+} | {
+    readonly result: "done";
+    readonly value: T;
 };
+export const raceTimer = async <T>(
+    p: Promise<T>,
+    timer: Promise<IPromiseTimerResult>,
+): Promise<RaceTimerResult<T>> => Promise.race([
+    p.then(
+        (value) => ({ result: "done", value } as const),
+        (error) => ({ result: "timeoutError", error } as const),
+    ),
+    timer.then(
+        (result) => ({ result: result.timerResult } as const),
+        (error) => ({ result: "error", error } as const),
```

Comment:
What are the cases when this error can happen? If this is not expected result (i.e. other than logic error in small amount of code, we do not expect to hit it), I'd rather use fail-fast approach and fail in some way that does not expose this kind of path to consumers. I.e. either log and ignore, or roll it into error path for p, or make it unhandled promise - something that the caller does not need to think about.

Reply:
You can actually see it from usage - there is only checking for "done" state, the rest of the code assumes that not done means timeout. Which is not entirely correct, as likely we want to differentiate error (not timer error) from timeout

Reply:
Right, I was just trying to propagate max info out of this function and let caller decide what to do- fail fast. I'll rethink this.

Reply:
I added different errors.. let me know if you think they are any better.

Reply:
Okay I have left them as promise rejections now. This way they will be sweeped up by parent catches and logged as "unexpected" errors in the corresponding places: SummaryManager and RunningSummarizer.

Reply:
@vladsud - bump!